OCPBUGS-26172 Configuring the master interface in the container network namespace with master bridge interface (#69965)
openshift/openshift-docs: modules/nw-multus-create-master-interface-bridge-cni.adoc

.Procedure

. Create a dedicated container namespace where you want to deploy your pod by running the following command:
+
[source,terminal]
----
$ oc new-project test-namespace
----

. Create a bridge network attachment definition file named `bridge-nad.yaml` and paste the following content:
+
[source,yaml]
----
apiVersion: "k8s.cni.cncf.io/v1"
kind: NetworkAttachmentDefinition
metadata:
  name: bridge-network
spec:
  config: '{
    "cniVersion": "0.4.0",
    "name": "bridge-network",
    "type": "bridge",
    "bridge": "br-001",
    "isGateway": true,
    "ipMasq": true,
    "hairpinMode": true,
    "ipam": {
      "type": "host-local",
      "subnet": "10.0.0.0/24",
      "routes": [{"dst": "0.0.0.0/0"}]
    }
  }'
----

. Run the following command to apply the NetworkAttachmentDefinition to your OpenShift cluster:
+
[source,terminal]
----
$ oc apply -f bridge-nad.yaml
----

. Verify that the NetworkAttachmentDefinition has been created successfully by running the following command:
+
[source,terminal]
----
$ oc get network-attachment-definitions
----
+
.Expected output

[source,terminal]
----
NAME             AGE
bridge-network   15s
----

> . Create a YAML file for the IPVLAN additional network configuration by saving the following YAML in a file named `ipvlan100-additional-network-configuration.yaml`:

Reviewer: remove the 100

    "master": "ext0", <1>
    "mode": "l3",
    "linkInContainer": true, <2>
    "ipam": {"type": "whereabouts", "ipRanges": [{"range": "10.0.0.0/24"}]}
  }'
----
+
<1> Specifies the Ethernet interface to associate with the network attachment. This is subsequently configured in the pod networks annotation.
<2> Specifies that the master interface is in the container network namespace.

. Apply the YAML by running the following command:
+
[source,terminal]
----
$ oc apply -f ipvlan100-additional-network-configuration.yaml
----

. Verify that the NetworkAttachmentDefinition has been created successfully by running the following command:
+
[source,terminal]
----
$ oc get network-attachment-definitions
----
+
.Expected output

[source,terminal]
----
NAME             AGE
bridge-network   9m36s
ipvlan-100       9s
----

. Create a pod definition by using the following YAML. Save the YAML in a file named `pod-a.yaml`.
+
[source,yaml]
----
apiVersion: v1
kind: Pod
metadata:
  name: pod-a
  namespace: test-namespace
  annotations:
    k8s.v1.cni.cncf.io/networks: '[
      {
        "name": "bridge-network",
        "interface": "ext0" <1>
      },
      {
        "name": "ipvlan-100",

>         "interface": "ext0.100"

Reviewer: use another name. This gives the impression that the interface has vlan 100

Author: so something like ipvlan-network would do @mlguerrero12

Reviewer: it could be
name: ipvlan-net
interface: ext1

// Module included in the following assemblies:
//
// * networking/multiple_networks/configuring-additional-network.adoc

:_mod-docs-content-type: PROCEDURE
[id="nw-multus-create-master-interface-bridge-cni_{context}"]
= Creating a subinterface based on a bridge master interface in a container namespace

Creating a subinterface can be applied to other types of interfaces. Follow this procedure to create a subinterface based on a bridge master interface in a container namespace.

.Prerequisites

> * You have installed the {product-title} CLI (oc).

Reviewer 2: Not critical, but we write this as:

```suggestion
* You have installed the OpenShift CLI (`oc`).
```

* You are logged in to the {product-title} cluster as a user with `cluster-admin` privileges.

.Procedure

. Create a dedicated container namespace where you want to deploy your pod by running the following command:
+
[source,terminal]
----
$ oc new-project test-namespace
----

. Using the following YAML example, create a bridge `NetworkAttachmentDefinition` custom resource (CR) file named `bridge-nad.yaml`:
+
[source,yaml]
----
apiVersion: "k8s.cni.cncf.io/v1"
kind: NetworkAttachmentDefinition
metadata:
  name: bridge-network
spec:
  config: '{
    "cniVersion": "0.4.0",
    "name": "bridge-network",
    "type": "bridge",
    "bridge": "br-001",
    "isGateway": true,
    "ipMasq": true,
    "hairpinMode": true,
    "ipam": {
      "type": "host-local",
      "subnet": "10.0.0.0/24",
      "routes": [{"dst": "0.0.0.0/0"}]
    }
  }'
----

> . Run the following command to apply the `NetworkAttachmentDefinition` CR to your OpenShift cluster:

Reviewer 2: ```suggestion
. Run the following command to apply the `NetworkAttachmentDefinition` CR to your {product-title} cluster:
```

+
[source,terminal]
----
$ oc apply -f bridge-nad.yaml
----

.  Verify that the `NetworkAttachmentDefinition` CR has been created successfully by running the following command:
+
[source,terminal]
----
$ oc get network-attachment-definitions
----
+

> .Expected output

Reviewer 2: We use "Example output" consistently
```suggestion
.Example output
```

Author: I thought Expected output would apply in this case as this is a worked example end-to-end and this is the expected output based on this worked example. It is not example output as such.

[source,terminal]
----
NAME             AGE
bridge-network   15s
----

. Using the following YAML example, create a file named `ipvlan-additional-network-configuration.yaml` for the IPVLAN additional network configuration:
+
[source,yaml]
----
apiVersion: k8s.cni.cncf.io/v1
kind: NetworkAttachmentDefinition
metadata:
  name: ipvlan-net
  namespace: test-namespace
spec:
  config: '{
    "cniVersion": "0.3.1",
    "name": "ipvlan-net",
    "type": "ipvlan",
    "master": "ext0", <1>
    "mode": "l3",
    "linkInContainer": true, <2>
    "ipam": {"type": "whereabouts", "ipRanges": [{"range": "10.0.0.0/24"}]}
  }'
----
+
<1> Specifies the ethernet interface to associate with the network attachment. This is subsequently configured in the pod networks annotation.
<2> Specifies that the master interface is in the container network namespace.

. Apply the YAML file by running the following command:
+
[source,terminal]
----
$ oc apply -f ipvlan-additional-network-configuration.yaml
----

. Verify that the `NetworkAttachmentDefinition` CR has been created successfully by running the following command:
+
[source,terminal]
----
$ oc get network-attachment-definitions
----
+

> .Expected output

Reviewer 2: ```suggestion
.Example output
```

[source,terminal]
----
NAME             AGE
bridge-network   87s
ipvlan-net       9s
----

. Using the following YAML example, create a file named `pod-a.yaml` for the pod definition:
+
[source,yaml]
----
apiVersion: v1
kind: Pod
metadata:
  name: pod-a
  namespace: test-namespace
  annotations:
    k8s.v1.cni.cncf.io/networks: '[
      {
        "name": "bridge-network",
        "interface": "ext0" <1>
      },
      {
        "name": "ipvlan-net",
        "interface": "ext1"
      }
    ]'
spec:
  securityContext:
    runAsNonRoot: true
    seccompProfile:
      type: RuntimeDefault
  containers:
  - name: test-pod
    image: quay.io/openshifttest/hello-sdn@sha256:c89445416459e7adea9a5a416b3365ed3d74f2491beb904d61dc8d1eb89a72a4
    securityContext:
      allowPrivilegeEscalation: false
      capabilities:
        drop: [ALL]
----
+
<1> Specifies the name to be used as the master for the IPVLAN interface.

. Apply the YAML file by running the following command:
+
[source,terminal]
----
$ oc apply -f pod-a.yaml
----

. Verify that the pod is running by using the following command:
+
[source,terminal]
----
$ oc get pod -n test-namespace
----
+

> .Expected output

Reviewer 2: ```suggestion
.Example output
```

// Module included in the following assemblies:
//

:_mod-docs-content-type: PROCEDURE
[id="nw-multus-create-master-interface-bridge-cni_{context}"]
= Creating a subinterface based on a bridge master interface in a container namespace

Creating a subinterface can be applied to other types of interfaces. Follow this procedure to create a subinterface based on a bridge master interface in a container namespace.

.Prerequisites
* You installed the OpenShift CLI (`oc`).
* You have access to the cluster as a user with the `cluster-admin` role.

.Procedure

. Create a dedicated container namespace where you want to deploy your pod by using the following command:
+
[source,terminal]
----
$ oc new-project test-namespace
----

. Create a bridge network attachment definition file named `bridge-nad.yaml` and paste the following content:
+
[source,yaml]
----
apiVersion: "k8s.cni.cncf.io/v1"
kind: NetworkAttachmentDefinition
metadata:
  name: bridge-network
spec:
  config: '{

>     "cniVersion": "0.3.1",

Reviewer 3: the `cniVersion` better keep same ipvlan 0.4.0
